src/radiation/mocmc.cpp

```diff
@@ -0,0 +1,1100 @@
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
+
+  auto geom = Geometry::GetCoordinateSystem(rc);
+
+  std::vector<std::string> variables{ir::tilPi, pf::velocity};
+  PackIndexMap imap;
+  auto v = rc->PackVariables(variables, imap);
+
+  const auto &Inuinv = swarm->template Get<Real>("Inuinv").Get();
+  const auto &ncov = swarm->template Get<Real>("ncov").Get();
+  const auto &mu_lo = swarm->template Get<Real>("mu_lo").Get();
+  const auto &mu_hi = swarm->template Get<Real>("mu_hi").Get();
+  const auto &phi_lo = swarm->template Get<Real>("phi_lo").Get();
+  const auto &phi_hi = swarm->template Get<Real>("phi_hi").Get();
+  auto iTilPi = imap.GetFlatIdx(ir::tilPi);
+  const auto pvel_lo = imap[pf::velocity].first;
+
+  const int nu_bins = rad->Param<int>("nu_bins");
+  const Real dlnu = rad->Param<Real>("dlnu");
+  auto nusamp = rad->Param<ParArray1D<Real>>("nusamp");
+  auto species = rad->Param<std::vector<RadiationType>>("species");
+  auto num_species = rad->Param<int>("num_species");
+  RadiationType species_d[3] = {};
+  for (int s = 0; s < num_species; s++) {
+    species_d[s] = species[s];
+  }
+  constexpr int MAX_SPECIES = 3;
+
+  // TODO(BRR) block packing eventually
+  const int iblock = 0;
+
+  auto swarm_d = swarm->GetDeviceContext();
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "MOCMC::kdgrid", DevExecSpace(), kb.s, kb.e, jb.s, jb.e, ib.s,
+      ib.e, KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        // initialize eddington to zero
+        for (int s = 0; s < num_species; s++) {
+          for (int ii = 0; ii < 3; ii++) {
+            for (int jj = ii; jj < 3; jj++) {
+              v(iTilPi(s, ii, jj), k, j, i) = 0.0;
+            }
+          }
+        }
+        Real energy[MAX_SPECIES] = {0.0};
+        const int nsamp = swarm_d.GetParticleCountPerCell(k, j, i);
+
+        Real cov_g[4][4];
+        geom.SpacetimeMetric(CellLocation::Cent, iblock, k, j, i, cov_g);
+        Real alpha = geom.Lapse(CellLocation::Cent, iblock, k, j, i);
+        Real con_beta[3];
+        geom.ContravariantShift(CellLocation::Cent, iblock, k, j, i, con_beta);
+        const Real vpcon[] = {v(pvel_lo, k, j, i), v(pvel_lo + 1, k, j, i),
+                              v(pvel_lo + 2, k, j, i)};
+        const Real W = phoebus::GetLorentzFactor(vpcon, cov_g);
+        const Real ucon[4] = {W / alpha, vpcon[0] - con_beta[0] * W / alpha,
+                              vpcon[1] - con_beta[1] * W / alpha,
+                              vpcon[2] - con_beta[2] * W / alpha};
+
+        for (int n = 0; n < nsamp; n++) {
+          const int nswarm = swarm_d.GetFullIndex(k, j, i, n);
+
+          Real nu_fluid0 = nusamp(0);
+          Real nu_lab0 = 0.;
+          SPACETIMELOOP(nu) { nu_lab0 -= ncov(nu, nswarm) * ucon[nu]; }
+          nu_lab0 *= nusamp(0);
+
+          const Real shift = (log(nusamp(0)) - log(nu_lab0)) / dlnu;
+          interpolation::PiecewiseConstant interp(nu_bins, dlnu, shift);
```

Comment:
Really feels like we want at least piecewise linear here.

Reply:
Yeah piecewise constant is yucky. I remember having some issues previously with scattering where the interpolations to the fluid frame and then back to the coordinate frame weren't exactly cancelling, and so that disagreement became a larger error in the sample spectra relative to the actual problem evolution. Something to look into in the (near) future.

Reply:
:+1: